diff --git a/src/os/exec/exec.go b/src/os/exec/exec.go
--- a/src/os/exec/exec.go
+++ b/src/os/exec/exec.go
@@ -633,6 +633,10 @@ func (c *Cmd) Run() error {
 // After a successful call to Start the [Cmd.Wait] method must be called in
 // order to release associated system resources.
 func (c *Cmd) Start() error {
+	if runtime.GOOS == "wasip1" {
+		return errors.New("exec: process execution is unavailable on wasip1")
+	}
+
 	// Check for doubled Start calls before we defer failure cleanup. If the prior
 	// call to Start succeeded, we don't want to spuriously close its pipes.
 	if c.Process != nil {
