diff --git a/Sources/_SwiftSyntaxCShims/include/_includes.h b/Sources/_SwiftSyntaxCShims/include/_includes.h
--- a/Sources/_SwiftSyntaxCShims/include/_includes.h
+++ b/Sources/_SwiftSyntaxCShims/include/_includes.h
@@ -31,7 +31,7 @@
 // swiftWinSDK.dll when (re)bulding it, and fails because it's used.
 #include <io.h>
 
-#elif defined(__unix__) || defined(__APPLE__)
+#elif defined(__unix__) || defined(__APPLE__) || defined(__wasi__)
 #include <unistd.h>
 #endif
 
