diff --git a/lldb/source/Host/CMakeLists.txt b/lldb/source/Host/CMakeLists.txt
--- a/lldb/source/Host/CMakeLists.txt
+++ b/lldb/source/Host/CMakeLists.txt
@@ -150,6 +150,16 @@ else()
   endif()
 endif()
 
+if(EMSCRIPTEN)
+  list(REMOVE_ITEM HOST_SOURCES
+    posix/ProcessLauncherPosixFork.cpp
+    )
+  list(APPEND HOST_SOURCES
+    emscripten/HostEmscripten.cpp
+    emscripten/HostInfoEmscripten.cpp
+    )
+endif()
+
 set(EXTRA_LIBS)
 if (CMAKE_SYSTEM_NAME MATCHES "NetBSD")
   list(APPEND EXTRA_LIBS kvm)
diff --git a/lldb/source/Host/common/Host.cpp b/lldb/source/Host/common/Host.cpp
--- a/lldb/source/Host/common/Host.cpp
+++ b/lldb/source/Host/common/Host.cpp
@@ -63,7 +63,7 @@
 #if defined(_WIN32)
 #include "lldb/Host/windows/ConnectionGenericFileWindows.h"
 #include "lldb/Host/windows/ProcessLauncherWindows.h"
-#else
+#elif !defined(__EMSCRIPTEN__)
 #include "lldb/Host/posix/ProcessLauncherPosixFork.h"
 #endif
 
@@ -535,6 +535,11 @@ Status Host::RunShellCommand(llvm::StringRef shell_path, const Args &args,
 // platforms
 #if !defined(__APPLE__)
 Status Host::LaunchProcess(ProcessLaunchInfo &launch_info) {
+#if defined(__EMSCRIPTEN__)
+  (void)launch_info;
+  return Status::FromErrorString(
+      "launching native processes is unavailable in the browser");
+#else
   std::unique_ptr<ProcessLauncher> delegate_launcher;
 #if defined(_WIN32)
   delegate_launcher.reset(new ProcessLauncherWindows());
@@ -551,5 +556,6 @@ Status Host::LaunchProcess(ProcessLaunchInfo &launch_info) {
   launch_info.SetProcessID(process.GetProcessId());
 
   return error;
+#endif
 }
 #endif // !defined(__APPLE__)
